File: Reqnroll/Chapter15/15.6-PolymorphicDriverPattern/After/WIMP.Specs/Support/TestActionResult.cs
namespace WIMP.Specs.Support;

public record TestActionResult<TResult>
{
    public static readonly TestActionResult<TResult> NotExecuted =
        new(false, null, default!);

    public static TestActionResult<TResult> CreateSucceeded(
        TResult result = default!) => new(true, null, result);

    public static TestActionResult<TResult> CreateFailed(
        TestActionFailedException error) => new(false, error, default!);

    public bool WasExecuted => !Equals(NotExecuted);

    public bool Success { get; }

    public TestActionFailedException? Error { get; }

    public TResult Result { get; }

    private TestActionResult(bool success, TestActionFailedException? error, TResult result)
    {
        Success = success;
        Error = error;
        Result = result;
    }

    public void AssertExecuted()
    {
        Assert.IsTrue(WasExecuted, "The action was not executed");
    }

    public void AssertFailed()
    {
        AssertExecuted();
        Assert.IsFalse(Success, $"The result ({this}) expected to be failure");
        Assert.IsNotNull(Error, $"The result ({this}) expected to contain an error");
    }

    public void AssertFailedWithErrorMessageContains(string expectedErrorMessage)
    {
        AssertFailed();
        StringAssert.Contains(Error!.Message, expectedErrorMessage);
    }

    public void AssertSucceeded()
    {
        AssertExecuted();
        Assert.IsTrue(Success, $"The result ({this}) expected to succeed");
    }
}

File: CucumberJVM/Chapter19/19.3-TestLogPattern/Before/src/test/java/com/wimp/app/specs/support/TestActionResult.java
package com.wimp.app.specs.support;

import static org.junit.jupiter.api.Assertions.*;

public class TestActionResult<TResult> {

    public static <TResult> TestActionResult<TResult> notExecuted() {
        return new TestActionResult<>(false, false, null, null);
    }

    public static <TResult> TestActionResult<TResult> createSucceeded(TResult result) {
        return new TestActionResult<>(true, true, null, result);
    }

    public static <TResult> TestActionResult<TResult> createSucceeded() {
        return createSucceeded(null);
    }

    public static <TResult> TestActionResult<TResult> createFailed(TestActionFailedException error) {
        return new TestActionResult<>(true, false, error, null);
    }

    private final boolean executed;
    private final boolean success;
    private final TestActionFailedException error;
    private final TResult result;

    public boolean wasExecuted() {
        return executed;
    }

    public boolean isSuccess() {
        return success;
    }

    public TestActionFailedException getError() {
        return error;
    }

    public TResult getResult() {
        return result;
    }

    private TestActionResult(boolean executed, boolean success, TestActionFailedException error, TResult result) {
        this.executed = executed;
        this.success = success;
        this.error = error;
        this.result = result;
    }

    public void assertExecuted() {
        assertTrue(wasExecuted(), "The action was not executed");
    }

    public void assertFailed() {
        assertExecuted();
        assertFalse(isSuccess(), "The result (%s) expected to be failure".formatted(this));
        assertNotNull(getError(), "The result (%s) expected to contain an error".formatted(this));
    }

    public void assertFailedWithErrorMessageContains(String expectedErrorMessage) {
        assertFailed();
        assertTrue(getError().getMessage().contains(expectedErrorMessage), "The error `%s` is expected to contain message `%s`".formatted(getError().getMessage(), expectedErrorMessage));
    }

    public void assertSucceeded() {
        assertExecuted();
        assertTrue(isSuccess(), "The result (%s) expected to succeed".formatted(this));
    }
}
